src/hb-ot-layout.cc

```diff
@@ -1949,7 +1962,89 @@ hb_ot_layout_get_baseline (hb_font_t               *font,
 			   hb_tag_t                 language_tag,
 			   hb_position_t           *coord        /* OUT.  May be NULL. */)
 {
-  return font->face->table.BASE->get_baseline (font, baseline_tag, direction, script_tag, language_tag, coord);
+#define INVALID INT_MAX
+  *coord = INVALID;
+  bool result = font->face->table.BASE->get_baseline (font, baseline_tag, direction, script_tag, language_tag, coord);
+
+  if (result && coord) *coord = font->em_scale_dir (*coord, direction);
+
+  constexpr hb_ot_layout_baseline_t ideo_tag = HB_OT_LAYOUT_BASELINE_IDEOGRAPHIC_EMBOX_BOTTOM_OR_LEFT_EDGE;
+  constexpr hb_ot_layout_baseline_t idtp_tag = HB_OT_LAYOUT_BASELINE_IDEOGRAPHIC_EMBOX_TOP_OR_RIGHT_EDGE;
+  constexpr hb_ot_layout_baseline_t icfb_tag = HB_OT_LAYOUT_BASELINE_IDEOGRAPHIC_CHARACTER_FACE_BOTTOM_OR_LEFT_EDGE;
+  constexpr hb_ot_layout_baseline_t icft_tag = HB_OT_LAYOUT_BASELINE_IDEOGRAPHIC_CHARACTER_FACE_TOP_OR_RIGHT_EDGE;
+
+  if (!coord || result ||
+      (baseline_tag != ideo_tag && baseline_tag != idtp_tag &&
+       baseline_tag != icfb_tag && baseline_tag != icft_tag)) /* tags we are able to simulate */
+    return result;
+
+  /* Simulate */
+  unsigned int upem = font->face->get_upem ();
+  bool is_horizontal = HB_DIRECTION_IS_HORIZONTAL (direction);
+  hb_position_t ideoEmboxBottom = INVALID, ideoEmboxLeft = INVALID, ideoEmboxTop = INVALID, ideoEmboxRight = INVALID;
+  hb_position_t icfBottom = INVALID, icfLeft = INVALID, icfTop = INVALID, icfRight = INVALID;
+  struct { hb_position_t ideo = INVALID, idtp = INVALID, icfb = INVALID, icft = INVALID; } HorizAxis, VertAxis;
+  hb_position_t margin = INVALID;
+#define GET_BASE(axis, member) font->face->table.BASE->get_baseline (font, member##_tag, \
+				&axis == &HorizAxis ? HB_DIRECTION_TTB : HB_DIRECTION_LTR, \
+				script_tag, language_tag, &axis.member)
+
+  /* https://docs.microsoft.com/en-us/typography/opentype/spec/baselinetags#ideographic-em-box */
+  ideoEmboxLeft = 0;
+  if (GET_BASE (HorizAxis, ideo))
+  {
+    ideoEmboxBottom = HorizAxis.ideo;
+    if (GET_BASE (HorizAxis, idtp))
+      ideoEmboxTop = HorizAxis.idtp;
+    else
+      ideoEmboxTop = HorizAxis.ideo + upem;
+  }
+  if (GET_BASE (VertAxis, idtp))
+    ideoEmboxRight = VertAxis.idtp;
+  else
+    ideoEmboxRight = upem;
+  if (GET_BASE (VertAxis, ideo) && VertAxis.ideo)
+  { /* Warning: Bad VertAxis.ideo value */ }
+  else /* If this is a CJK font: */ /* But we haven't implemented this check */
+  {
+    ideoEmboxBottom = font->face->table.OS2->sTypoDescender;
+    ideoEmboxTop = font->face->table.OS2->sTypoAscender;
```

Comment:
I couldn't use the just introduced ot-metrics API as the it is needed to have the value unscaled till the end.

Reply:
Umm. Need to apply the sign conversion.  Plus, variations.  I think you have to use the metrics API.  Again, remove the fallback and we rebuild it separately.

Reply:
Removed simulation and put it here for future reference https://github.com/harfbuzz/harfbuzz/compare/master...ebraminio:baseline-base